File: typescript/session3/td/UnionRec.ts
import { UnionIter } from './UnionIter';
import { ConsIter } from './ConsIter';
import { Mot } from "./Mot"
import { Iterateur } from './Iterateur';

export class UnionRec extends Mot {
    public motGauche : Mot;
    public motDroit : Mot;
    public var_taille : number;

    public constructor(motGauche : Mot, motDroit : Mot) {
        super();
        this.motGauche = motGauche;
        this.motDroit = motDroit;
        this.var_taille = motGauche.taille() + motDroit.taille();
    }

    public casUnion() : boolean {
        return true;
    }

    public gauche() : Mot {
        return this.motGauche;
    }

    public droit() : Mot {
        return this.motDroit;
    }

    public taille() : number {
        return this.var_taille;
    }

    public lettre() : string {
        if(!this.motGauche.estVide()) {
			return this.motGauche.lettre();
        } else if(!this.motDroit.estVide()) {
            return this.motDroit.lettre();
        } else {
            throw Error("Unsupported operation");
        }	
    }

    public reste(): Mot {
        if(!this.motGauche.estVide()) {
			return new UnionIter(this.motGauche.reste(), this.motDroit);
        } else if(!this.motDroit.estVide()) {
			return this.motDroit.reste();
        } else {
            throw Error("Unsupported operation");
        }
    }
}

File: java/cours/src/session3/td/src/UnionRec.java
public class UnionRec implements Mot{
    private Mot gauche;
    private Mot droit;
    private int taille;

    UnionRec(Mot gauche, Mot droit) {
        this.gauche = gauche;
        this.droit = droit;
        this.taille = gauche.taille() + droit.taille();
    }

    @Override
    public Mot gauche() {
        return gauche;
    }

    @Override
    public Mot droit() {
        return droit;
    }

    @Override
    public boolean casUnion() {
        return true;
    }

    @Override
    public int taille() {
        return taille;
    }

    public String toString() {
        return gauche.toString() + droit.toString();
    }

    @Override
    public char lettre() {
        if(!gauche.estVide())
            return this.gauche.lettre();
        if(!droit.estVide())
            return this.droit.lettre();
        throw new UnsupportedOperationException();
    }

    @Override
    public Mot reste() {
        if(!gauche.estVide())
            return new UnionRec(this.gauche.reste(), this.droit);
        if(!droit.estVide())
            return this.droit.reste();
        throw new UnsupportedOperationException();
    }
}
